In [19]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [20]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
url = './science_data_large.csv'
df = pd.read_csv(url)
print(df.head(15))
print(df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [21]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = df.drop(columns=['Size nm^3'])
y = df['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

## Part 3. Perform a Linear Regression

In [22]:
# Use sklearn to train a model on the training set
model = LinearRegression()
# Create a sample datapoint and predict the output of that sample with the trained model
model.fit(X_train, y_train)
sample = df.iloc[0].drop('Size nm^3').values.reshape(1, -1)
print(sample)
#TODO: ask whether i should use the test set or the training set for the sample
prediction = model.predict(sample)
print('Prediction:', prediction)
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
modelScore = model.score(X_test, y_test)
print('Model Score:', modelScore)
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
cofficents = model.coef_
intercept = model.intercept_
print('Coefficients:', cofficents)
print('Weight of feature one:', cofficents[0])
print('Weight of feature two:', cofficents[1])
print('Intercept:', intercept)

[[469. 647.]]
Prediction: [664984.89630433]
Model Score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Weight of feature one: 866.1464133719209
Weight of feature two: 1032.6950664857964
Intercept: -409391.47958340764


/home/codespace/.local/lib/python3.12/site-packages/sklearn/base.py:493: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


## Report on the score of the model
The $(R^2)$ score of **0.8552** indicates that the regression line approximates the actual data quite well. This score means that about **85.52%** of the variance in the actual data is captured by the regression line, leaving roughly **14.48%** of the variance unexplained. In practical terms, the regression line fits closely to the data points, but it doesn't perfectly match all of them. The higher the $(R^2)$ value, the better the line represents the actual data, so this value shows a strong, but not perfect, approximation.

## Equation of linear regression:
$x_1$ = Temperature , $x_2$ = Mols KCl

$h(x) = 866.1464 \cdot x_1 + 1032.6951 \cdot x_2 −409,391.4796$

## Part 4. Use Cross Validation

In [23]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
scores = cross_val_score(model, X, y, cv=10)
print('Cross Validation Scores:', scores)
print('The mean/average of the cross validation scores: ', scores.mean())
# Report on their finding and their significance

Cross Validation Scores: [0.81123596 0.86440978 0.87808742 0.86561069 0.87495621 0.84484397
 0.87941022 0.86349411 0.78353682 0.88686516]
The mean/average of the cross validation scores:  0.8552450341984701


## Findings and significance of the cross-validation results:

The cross-validation scores obtained from the model represent the $R^2$ values for each of the 10 folds. These scores demonstrate that the model performs consistently across different portions of the dataset, with values ranging between **0.811** and **0.887**. This consistency indicates that the model is reliable and also might not overfit to any particular subset of data. The average score of **0.855** closely aligns with the previously reported $R^2$ score of **0.8552**, confirming that the model generalizes well and maintains stable performance. The small variance in scores further suggests that the model is robust and not overly sensitive to how the data is split for training and testing.

## Part 5. Using Polynomial Regression

In [24]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
Poly = PolynomialFeatures(degree=2)
X_poly = Poly.fit_transform(X)
X_train_poly, X_test_poly, y_train, y_test = train_test_split(X_poly, y, test_size=0.1, random_state=42)
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train)
y_pred = poly_model.predict(X_test_poly)
poly_model_score = poly_model.score(X_test_poly, y_test)
print('Poly Model Score:', poly_model_score)
poly_cofficients = poly_model.coef_
poly_intercept = poly_model.intercept_
print('Poly Coefficients:', poly_cofficients)
print('Poly Intercept:', poly_intercept)
# Report on the metrics and output the resultant equation as you did in Part 3.

Poly Model Score: 1.0
Poly Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Poly Intercept: 2.0477105863392353e-05


## Report on the findings of the polynomial regression model:

Given the note about ignoring 'values outside of 5 significant figures'
The updated coefficients and intercept for the polynomial regression model are as follows:

Coefficients: [0, 12.0000, 0, 0, 2, 0.028571]

From documentation: [1, a, b, a^2, ab, b^2]

Intercept: 0

$h(x) = 12 \cdot a + 2 \cdot ab + 0.028571 \cdot b^2$


The polynomial regression model I built got a perfect \(R^2\) score of **1.0**, meaning it can explain 100% of the differences in the target variable based on the input features. This likely means the model is **overfitting**. Overfitting happens when the model learns not just the actual patterns in the data, but also the random noise or irrelevant details. This can make the model perform very well on the current dataset but struggle when making predictions on new data. In other words, the model may not be able to generalize well to unseen data, and its perfect performance on this set could be misleading.